### PR TITLE
[AutoModel] Add AutoModelForTextEncoding

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Adds AutoModel for text encoding (used in the circumstance when you want to extract the text encoder from an encoder-decoder architecture).

This facilitates loading a t5 encoder from t5 enc-dec model weights (as is done in Music Gen in #24109)

### CODE REVIEW

File: src/transformers/models/auto/modeling_auto.py [hunk 1]
```diff
@@ -1011,6 +1011,36 @@
     ]
 )
 
+MODEL_FOR_TEXT_ENCODING_MAPPING_NAMES = OrderedDict(
```
Comment (sanchit-gandhi) [inline src/transformers/models/auto/modeling_auto.py:1014]: Is this list good for you @amyeroberts? It's the most popular encoder-only architectures, and enc-dec ones where we have the encoder model separated as an import

Comment (amyeroberts) [reply src/transformers/models/auto/modeling_auto.py:1014]: Yep! We can always add as needed :)

File: src/transformers/utils/dummy_pt_objects.py [hunk 2]
```diff
@@ -716,6 +716,13 @@ def __init__(self, *args, **kwargs):
         requires_backends(self, ["torch"])
 
 
+class AutoModelForTextEncoding(metaclass=DummyObject):
```
Comment (amyeroberts) [inline src/transformers/utils/dummy_pt_objects.py:719]: There should be the equivalent `TFAutoModelForTextEncoding` in `dummy_tf_objects.py`

File: src/transformers/__init__.py [hunk 3]
```diff
@@ -1084,6 +1084,7 @@
             "AutoModelForSequenceClassification",
             "AutoModelForSpeechSeq2Seq",
             "AutoModelForTableQuestionAnswering",
+            "AutoModelForTextEncoding",
```
Comment (amyeroberts) [inline src/transformers/__init__.py:1087]: The TF equivalent class should be in this init too

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! 

Just a comment on some potential missing imports for TF model in the init

File: src/transformers/utils/dummy_pt_objects.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/utils/dummy_pt_objects.py:719]: Ah perfect! Thanks for flagging!

File: src/transformers/__init__.py (hunk 3 above)
Comment (sanchit-gandhi) [reply src/transformers/__init__.py:1087]: Resolved in https://github.com/huggingface/transformers/pull/24305/commits/8b003f7d01556b0424550c784149f7d775989e3e

**Pull request merged**